clusterloader2/pkg/measurement/common/api_availability_measurement.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"k8s.io/perf-tests/clusterloader2/pkg/errors"
+	"sync"
+	"time"
+
+	goerrors "github.com/go-errors/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement"
+	measurementutil "k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+)
+
+const (
+	apiAvailabilityName = "APIAvailability"
+)
+
+func init() {
+	if err := measurement.Register(apiAvailabilityName, createAPIAvailabilityMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", apiAvailabilityName, err)
+	}
+}
+
+func createAPIAvailabilityMeasurement() measurement.Measurement {
+	return &apiAvailabilityMeasurement{}
+}
+
+type apiAvailabilityMeasurement struct {
+	isRunning           bool
+	stopCh              chan struct{}
+	hosts               []string
+	summaries           []measurement.Summary
+	clusterLevelMetrics *apiAvailabilityMetrics
+	hostLevelMetrics    map[string]*apiAvailabilityMetrics
+	wg                  sync.WaitGroup
+}
+
+func (a *apiAvailabilityMeasurement) updateMasterAvailabilityMetrics(c clientset.Interface, config *measurement.Config) {
+	for _, host := range a.hosts {
+		// SSH and check the health of the host
+		command := fmt.Sprintf("curl -s -k %slocalhost:%v/healthz", "https://", 443)
+		sshResult, err := measurementutil.SSH(command, host+":22", "")
+		if err != nil || sshResult.Code != 0 {
```

Comment:
What does curl return, if the server return say 500?

Reply:
I think it would be part of the stdout. So, might want to check for the status. I am not totally sure of the behaviour of the ssh session which executes the command. Is there a way to test?

Reply:
You can start locally a server that will always respond 500 and call this code against it.

Reply:
I was trying to write a test for this with the same ssh code. I have docker for desktop running on my local and wanted to ssh into a locally running kube-api server. The master IP that I am passing is kubernetes.docker.internal. I see that in the configs, there is an option to pass a MasterIP. But, on a locally running kube cluster, masterIP is local, right?

Note that, I am able to run the /healthz end points by setting up a clientset with the same kube config file and the below example you provided works.

Reply:
@wojtek-t I was little jammed with work this week but tried some more things today. I set up a docker container running a Flask application inside and deployed it through kubernetes. i also set up a mechanism to ssh into the container. 

Then why I ran it through the test case which invokes this logic, i found that any error in web app is returned in sshResult.stderr. I tried with 404 and got a big html like string (perhaps flask returns that). Note that sshResult.stdout is empty.

In the happy scenario, the sshResult.stdout has the response and sshResult.stderr is empty. 

So, to answer the original question, can I also make this assumption that the presence of stderr indicates the /healthz end point failed and vice versa in the other case? Should I also look to parse the stdout result?

PS Unfortunately, I still couldn't find a way to ssh into the kube-apiserver container. Otherwise I could have validated directly.

Reply:
hey @wojtek-t i am presuming you're busy... Just reaching out to ask did you get a chance to look at the comments?

Reply:
Sorry for delay.
It would be better to check it directly, but I don't want to block the whole PR on this thing.
So if you could open an issue to check the semantics of ssh and document it better and here in the code add a TODO (linked to this issue) to validate if the logic of relying on stderr non-empty is ok, I'm fine with that (resolving it should be a blocker for enabling the measurement though, but not necessary for merging this PR).

Reply:
Thanks @wojtek-t . As suggested, created a new issue to understand ssh client semantics:
https://github.com/kubernetes/perf-tests/issues/1343 and added a TODO here.
Just one thing, you asked me to document it you meant in the newly created issue?

Reply:
hey @wojtek-t I have the made the changes to this PR based upon my findings mentioned here:
https://github.com/kubernetes/perf-tests/issues/1343

Plz review whenever you have the time :)